GlotPress/GlotPress-WP
# Translation errors

## Problem

We have some situations where we don't check some translations or if we check it, we return a warning, that can be dismissed by the translator, so the translators can add incorrect, and sometimes, dangerous translations. Some examples:
* [Incorrect placeholders](https://github.com/GlotPress/GlotPress/issues/1546). 
* [Incorrect date format](https://translate.wordpress.org/projects/wp/dev/admin/gl/default/?filters%5Bstatus%5D=either&filters%5Boriginal_id%5D=9144374&filters%5Btranslation_id%5D=70640124). 
* [Incorrect timezone format](https://translate.wordpress.org/projects/wp/dev/admin/gl/default/?filters%5Bstatus%5D=either&filters%5Boriginal_id%5D=40444&filters%5Btranslation_id%5D=105190736). 

Fixes https://github.com/GlotPress/GlotPress/issues/1546.



## Solution

This PR adds two classes, to be able to add some checks in the GlotPress core, and another ones in an external plugin, because sometimes the error checks are very specific to the project (e.g., WordPress core).

This PR adds in the core the [Incorrect placeholder check](https://github.com/GlotPress/GlotPress/issues/1546). 

The end users will receive an error, so they can't dismiss the message, so with this update we will be sure that we check some big translation problems and the translator can't dismiss the message.



## To-do


- [ ] Create an external plugin, at translate.w.org, to add some checks for the WordPress core:
    - Check the unexpected timezone.
    - Check the unexpected start of the week.
    - Check the unexpected date format (year, month, day).
    - ...

## Testing Instructions


Before start testing, please, compile the assets executing:

```
npm install
npm run build
```

This PR add some unit tests in the `tests/phpunit/testcases/test_builtin_errors.php` file. To execute only these tests, you cas run 

`phpunit --filter GP_Test_Builtin_Translation_Errors
`

To execute all the tests, you can run

`phpunit` 

To test the *incorrect placeholder check* in the UI, you can:
1. Open an original string in a project.
2.  Add the `Yo! We so great! 100% bug free! <a href="%s">Check it!</a>` translation.
3. Click on the "Add translation" or "Suggest" button, depending on your role.
4. You will get an error in the UI and you can't save the translation.

![image](https://github.com/GlotPress/GlotPress/assets/1667814/1f4c3287-e255-4516-9ae1-f2b3d12bb3df)


## Screenshots or screencast
